apps/api/src/app/events/services/storage-helper-service/storage-helper.service.ts

```diff
@@ -0,0 +1,51 @@
+import { Injectable } from '@nestjs/common';
+import { IAttachmentOptions } from '@novu/stateless';
+import { StorageService } from '../../../shared/services/storage/storage.service';
+
+@Injectable()
+export class StorageHelperService {
+  constructor(private storageService: StorageService) {}
+
+  async uploadAttachments(attachments?: IAttachmentOptions[]) {
+    if (!(Array.isArray(attachments) && attachments.length > 0)) {
+      return;
+    }
+
+    const promises = attachments.map(async (attachment) => {
+      if (attachment.file) {
+        await this.storageService.uploadFile(attachment.name, attachment.file, attachment.mime);
+      }
+    });
+    await Promise.all(promises);
+  }
+
+  async getAttachments(attachments?: IAttachmentOptions[]) {
+    if (!(Array.isArray(attachments) && attachments.length > 0)) {
+      return;
+    }
+
+    for (const attachment of attachments) {
+      try {
+        attachment.file = await this.storageService.getFile(attachment.name);
+      } catch (error) {
+        if (error.message === 'The specified key does not exist.') {
```

Comment:
Where can we check this is the error message we might receive from the storage provider? Just asking because checking by the message can be a brittle check and if we could check by error name, or error code for the different providers would make it a stronger implementation.
Probably we would need a downstream `getFile` implementation where we throw a controlled error defined by us to unify it.

Reply:
Hey, @p-fernandez Can you give me a hint on `how to implement the controlled exception` you're talking about?

Reply:
In these 3 places of `StorageService` where we get the file for the different cloud services we let the errors bubble up as they come from the cloud providers:
https://github.com/novuhq/novu/blob/1525ebcaf957a44df0982f796cd8887b9210ccfd/apps/api/src/app/shared/services/storage/storage.service.ts#L61
https://github.com/novuhq/novu/blob/1525ebcaf957a44df0982f796cd8887b9210ccfd/apps/api/src/app/shared/services/storage/storage.service.ts#L114
https://github.com/novuhq/novu/blob/1525ebcaf957a44df0982f796cd8887b9210ccfd/apps/api/src/app/shared/services/storage/storage.service.ts#L168

So my suggestion is to implement a controlled exception in those 3 methods. Something in the lines of:
```
 async getFile(key: string): Promise<Buffer> {
    try { 
       ...
       // specific cloud provider code
     } catch (error) {
        // Here we would check the right way of that provider to know that the key doesn't exist
        // to be able to control the exception.
        if (condition) {
            throw new NonExistingFileError();
        } else {
           throw error;
        }
     }
  }
 ```
 Being `NonExistingFileError` something along these lines:
 ```
 class NonExistingFileError extends Error {
   constructor() {
      // The generic message we might think is appropriate
      super('File not found for the key provided');
      this.name = 'NonExistingFileError';
   }
 }
 ```
 
 For Google Cloud, if I recall correctly the implementation could be something along these lines:
 ```
 async getFile(key: string): Promise<Buffer> {
    try { 
        const bucket = this.gcs.bucket(process.env.GCS_BUCKET_NAME);
        const fileObject = bucket.file(key);
        return (await fileObject.download())[0] as unknown as Buffer;
     } catch (error) {
        // In GCP errors they usually come with a code that is more reliable to use than text checking as it is more likely to change.
        if (error.code === 404) {
            throw new NonExistingFileError();
        } else {
           throw error;
        }
     }
  }
```
For the other cloud providers the checks could be different based on other details of the error received. Worth to check their APIs to know what would be the most consistent way. Usually if they have error codes, seems like the strongest path to go.

So then, upstream here, we could better manage this case by doing:
```
 for (const attachment of attachments) {
      try {
        attachment.file = await this.storageService.getFile(attachment.name);
      } catch (error) {
        if (error instanceof NonExistingFileError) {
          attachment.file = null;
        } else {
          throw error;
        }
      }
  }
```
as for all providers we would be covered as they would throw the same error in StorageService and in case the cloud provider implementation changes individually we just need to change the appropriate one.

Hope this clarifies my suggestion.